web/server/net_logger_test.go

```diff
@@ -118,190 +52,177 @@ func TestNetLoggerQueueOperations(t *testing.T) {
+
+func (ms *mockRobotService) Log(ctx context.Context, req *apppb.LogRequest) (*apppb.LogResponse, error) {
+	if ms.expectedID != req.Id {
+		return nil, fmt.Errorf("expected id %q but got %q", ms.expectedID, req.Id)
+	}
+	ms.logsMu.Lock()
+	defer ms.logsMu.Unlock()
+	if ms.logFailForSizeCount > 0 {
+		ms.logFailForSizeCount -= len(req.Logs)
+		return &apppb.LogResponse{}, errors.New("not right now")
+	}
+	ms.logs = append(ms.logs, req.Logs...)
+	ms.logBatches = append(ms.logBatches, req.Logs)
+	return &apppb.LogResponse{}, nil
+}
 
-	client := mockapppb.NewMockRobotServiceClient(ctrl)
+type serverForRobotLogger struct {
+	service     *mockRobotService
+	cloudConfig *config.Cloud
+	stop        func() error
+}
 
+func makeServerForRobotLogger(t *testing.T) serverForRobotLogger {
 	logger := golog.NewTestLogger(t)
-	cancelCtx, cancel := context.WithCancel(context.Background())
+	listener, err := net.Listen("tcp", "localhost:0")
+	test.That(t, err, test.ShouldBeNil)
+	rpcServer, err := rpc.NewServer(logger, rpc.WithUnauthenticated())
+	test.That(t, err, test.ShouldBeNil)
+
+	robotService := &mockRobotService{expectedID: "abc-123"}
+	test.That(t, rpcServer.RegisterServiceServer(
+		context.Background(),
+		&apppb.RobotService_ServiceDesc,
+		robotService,
+		apppb.RegisterRobotServiceHandlerFromEndpoint,
+	), test.ShouldBeNil)
 
+	go rpcServer.Serve(listener)
 	config := &config.Cloud{
-		AppAddress: "http://localhost:8080",
-		ID:         "abc-123",
+		AppAddress: fmt.Sprintf("http://%s", listener.Addr().String()),
+		ID:         robotService.expectedID,
 	}
+	return serverForRobotLogger{robotService, config, rpcServer.Stop}
+}
 
-	logWriter := &remoteLogWriterGRPC{
-		loggerWithoutNet: logger,
-		cfg:              config,
-		service:          client,
-	}
+func TestNetLoggerBatchWrites(t *testing.T) {
+	server := makeServerForRobotLogger(t)
 
-	nl := &netLogger{
-		hostname:     "hostname",
-		cancelCtx:    cancelCtx,
-		cancel:       cancel,
-		remoteWriter: logWriter,
-		maxQueueSize: 100,
-		logLevel:     zap.NewAtomicLevelAt(zap.InfoLevel),
-	}
+	logger := golog.NewTestLogger(t)
+	nl, err := newNetLogger(server.cloudConfig, logger, zap.NewAtomicLevelAt(zap.InfoLevel))
+	test.That(t, err, test.ShouldBeNil)
 
-	l := logger.Desugar()
-	l = l.WithOptions(zap.WrapCore(func(c zapcore.Core) zapcore.Core {
+	loggerWithNet := logger.Desugar()
+	loggerWithNet = loggerWithNet.WithOptions(zap.WrapCore(func(c zapcore.Core) zapcore.Core {
 		return zapcore.NewTee(c, nl)
 	}))
 
-	client.EXPECT().
-		Log(gomock.Any(), &expectedLogMatch{nLogs: 10, id: "abc-123"}).
-		Times(1).
-		Return(nil, errors.New("failure"))
-
-	for i := 0; i < 10; i++ {
-		l.Info("Some-info")
+	for i := 0; i < writeBatchSize+1; i++ {
+		loggerWithNet.Info("Some-info")
 	}
-	nl.Sync()
-
-	l.Info("New info")
-	client.EXPECT().
-		Log(gomock.Any(), &expectedLogMatch{nLogs: 11, id: "abc-123"}).
-		Times(1).
-		Return(&apppb.LogResponse{}, nil)
 
 	nl.Sync()
-
 	nl.Close()
-}
+	server.stop()
```

Comment:
As far as I know, stopping the underlying gRPC server is the best way for us to introduce a happens before relationship in go's memory model.